accounts-db/src/accounts_hash.rs

```diff
@@ -944,41 +939,75 @@ impl<'a> AccountsHasher<'a> {
-    // 1. handle zero lamport accounts
-    // 2. pick the highest slot or (slot = and highest version) of each pubkey
-    // 3. produce this output:
-    //   a. AccountHashesFile: individual account hashes in pubkey order
-    //   b. lamport sum
-    fn de_dup_accounts_in_parallel(
-        &self,
+    /// Return the working_set and max number of pubkeys for hash dedup.
+    fn initialize_dedup_working_set(
         sorted_data_by_pubkey: &[&[CalculateHashIntermediate]],
         pubkey_bin: usize,
         bins: usize,
+        binner: &PubkeyBinCalculator24,
         stats: &HashStats,
-    ) -> (AccountHashesFile, u64) {
-        let binner = PubkeyBinCalculator24::new(bins);
-
-        let len = sorted_data_by_pubkey.len();
-        let mut indexes = Vec::with_capacity(len);
-        let mut first_items = Vec::with_capacity(len);
-        // map from index of an item in first_items[] to index of the corresponding item in sorted_data_by_pubkey[]
-        // this will change as items in sorted_data_by_pubkey[] are exhausted
-        let mut first_item_to_pubkey_division = Vec::with_capacity(len);
-
-        // initialize 'first_items', which holds the current lowest item in each slot group
+    ) -> (Vec<(usize, usize)>, usize) {
+        // working_set hold the lowest items for each slot_group sorted by pubkey descending (min_key is the last)
+        let mut working_set: Vec<(usize, usize)> = Vec::default();
+
+        // initialize 'working_set', which holds the current lowest item in each slot group
+        // working_set should be initialized in reverse order of slot_groups. Later slot_groups are
+        // processed first. For each slot_group, if the lowest item for current slot group is
+        // already in working_set (i.e. inserted by a later slot group), the next lowest items
+        // in this slot group are searched and checked until either one that is not the
+        // working_set are found, which will then be inserted, or no next lowest items are found.
         let max_inclusive_num_pubkeys = sorted_data_by_pubkey
             .iter()
             .enumerate()
+            .rev()
             .map(|(i, hash_data)| {
                 let first_pubkey_in_bin =
-                    Self::find_first_pubkey_in_bin(hash_data, pubkey_bin, bins, &binner, stats);
+                    Self::find_first_pubkey_in_bin(hash_data, pubkey_bin, bins, binner, stats);
+
                 if let Some(first_pubkey_in_bin) = first_pubkey_in_bin {
-                    let k = hash_data[first_pubkey_in_bin].pubkey;
-                    first_items.push(k);
-                    first_item_to_pubkey_division.push(i);
-                    indexes.push(first_pubkey_in_bin);
+                    let mut key = Some(&hash_data[first_pubkey_in_bin].pubkey);
+                    let mut offset = first_pubkey_in_bin;
+
+                    while let Some(k) = key {
```

Comment:
what do you think of replacing this loop with something like:
```
let next= hash_data[first_pubkey_in_bin...];

            Self::add_next_item(
                &mut next,
                &mut new,
                &mut working_set,
                ...
            );
            if let Some(new) = new {
            working_set.insert(0, new);
            }
            ```

            The logic in here is complicated and mirrors what is in `add_next_item`.

Reply:
yeah. done.